### PR TITLE
Whitelist/integration

### PR DESCRIPTION
**Pull request by Hirama**
The complete workflow for use whitelists

### CODE REVIEW

File: src/integration-test/kotlin/integration/helper/IntegrationHelperUtil.kt [hunk 1]
```diff
@@ -211,7 +213,8 @@ class IntegrationHelperUtil {
     fun registerBtcAddress(irohaAccountName: String): String {
         val keypair = ModelCrypto().generateKeypair()
         preGenBtcAddress().fold({
-            btcRegistrationStrategy.register(irohaAccountName, keypair.publicKey().hex())
+            // TODO: remove whiteList stub
```
Comment (Alexey-N-Chernyshov) [inline src/integration-test/kotlin/integration/helper/IntegrationHelperUtil.kt]: remove todo

File: src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt [hunk 2]
```diff
@@ -78,13 +84,19 @@ class EthRefundStrategyImpl(
                     val destEthAddress = ""
                     logger.info { "Rollback case ($key, $value)" }
 
-                    EthRefund(destEthAddress, "mockCoinType", "10", request.irohaTx)
+                    val relayAddress = relayProvider.getRelays().get().filter {
+                        it.value == commands.transferAsset.srcAccountId
+                    }.keys.first()
+
+                    println(relayAddress)
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt]: remove println

File: src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt [hunk 3]
```diff
@@ -108,9 +120,14 @@ class EthRefundStrategyImpl(
                             { throw it }
                         )
 
+                    val relayAddress = relayProvider.getRelays().get().filter {
+                        it.value == commands.transferAsset.srcAccountId
+                    }.keys.first()
+                    println("From Iroha $relayAddress")
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt]: sap

File: src/main/kotlin/registration/IrohaAccountCreator.kt [hunk 4]
```diff
@@ -29,31 +30,49 @@ class IrohaAccountCreator(
...
             val domain = "notary"
-            IrohaTransaction(
-                creator,
-                getCurrentTime(),
-                1,
-                arrayListOf(
-                    // Create account
-                    IrohaCommand.CommandCreateAccount(
-                        userName, domain, pubkey
+            IrohaOrderedBatch(
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/registration/IrohaAccountCreator.kt]: add todo link to https://soramitsu.atlassian.net/browse/D3-415

File: src/main/kotlin/registration/RegistrationServiceEndpoint.kt [hunk 5]
```diff
@@ -53,15 +54,20 @@ class RegistrationServiceEndpoint(
         return Response(code, errorMsg)
     }
 
-    private fun onPostRegistration(name: String?, pubkey: String?): Response {
-        if (name == null && pubkey == null) {
-            return responseError(HttpStatusCode.BadRequest, "Parameters \"name\" and \"pubkey\" are not specified.")
+    private fun onPostRegistration(name: String?, whitelist: List<String>?, pubkey: String?): Response {
+        if (name == null && whitelist == null && pubkey == null) {
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/registration/RegistrationServiceEndpoint.kt]: please, rework error handling

File: src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt [hunk 6]
```diff
@@ -2,33 +2,70 @@ package registration.eth
...
+            ethRegistrationConfig.ethRelayRegistryAddress,
+            web3,
+            credentials,
+            BigInteger.valueOf(ethRegistrationConfig.ethereum.gasPrice),
+            BigInteger.valueOf(ethRegistrationConfig.ethereum.gasLimit)
+        )
+
+        val irohaAccountCreator = IrohaAccountCreator(irohaConsumer, notaryIrohaAccount, creator, "ethereum_wallet")
+
+        println("ADDRESS " + relayRegistry.contractAddress)
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt]: remove println

File: src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt [hunk 7]
```diff
@@ -2,33 +2,70 @@ package registration.eth
...
+        println("NETWORK " + web3.netListening().send().id)
+
         return ethFreeRelayProvider.getRelay()
             .flatMap { freeEthWallet ->
-                irohaAccountCreator.create(freeEthWallet, name, pubkey)
+                println("Select next relay " + freeEthWallet.toString())
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt]: sap

File: src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt [hunk 8]
```diff
@@ -2,33 +2,70 @@ package registration.eth
...
+
+        val credentials = WalletUtils.loadCredentials(
+            passwordConfig.credentialsPassword,
+            ethRegistrationConfig.ethereum.credentialsPath
+        )
+        val builder = OkHttpClient().newBuilder().authenticator(BasicAuthenticator(passwordConfig))!!
+        println("URL " + ethRegistrationConfig.ethereum.url)
+        val web3 = Web3j.build(HttpService(ethRegistrationConfig.ethereum.url, builder.build(), false))
+
+        val relayRegistry = RelayRegistry.load(
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt]: you can move relayRegister out of function scope

File: src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt [hunk 9]
```diff
@@ -2,33 +2,66 @@ package registration.eth
...
-    creator: String
+    ethRegistrationConfig: EthRegistrationConfig,
+    passwordConfig: EthereumPasswords,
+    val irohaConsumer: IrohaConsumer,
+    val notaryIrohaAccount: String,
+    val creator: String
 ) : RegistrationStrategy {
-    private val irohaAccountCreator = IrohaAccountCreator(irohaConsumer, notaryIrohaAccount, creator, "ethereum_wallet")
+
+    val credentials = WalletUtils.loadCredentials(
```
Comment (Alexey-N-Chernyshov) [inline src/main/kotlin/registration/eth/EthRegistrationStrategyImpl.kt]: private

File: src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt [hunk 10]
```diff
@@ -78,13 +84,17 @@ class EthRefundStrategyImpl(
                     val destEthAddress = ""
                     logger.info { "Rollback case ($key, $value)" }
 
-                    EthRefund(destEthAddress, "mockCoinType", "10", request.irohaTx)
+                    val relayAddress = relayProvider.getRelays().get().filter {
```
Comment (dolgopolovwork) [inline src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt]: I think get() is not appropriate function here. I'd rather use map() or fold() instead

File: src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt [hunk 11]
```diff
@@ -108,9 +118,11 @@ class EthRefundStrategyImpl(
                             { throw it }
                         )
 
+                    val relayAddress = relayProvider.getRelays().get().filter {
```
Comment (dolgopolovwork) [inline src/main/kotlin/notary/endpoint/eth/EthRefundStrategyImpl.kt]: The same

File: src/main/kotlin/registration/RegistrationServiceEndpoint.kt [hunk 12]
```diff
@@ -53,24 +54,28 @@ class RegistrationServiceEndpoint(
...
-                return Response(HttpStatusCode.OK, address)
-            },
-            { ex ->
-                logger.error("Cannot register client $name", ex)
-                // TODO - D3-121 - a.chernyshov - distinguish correct status code response (500 - server internal error)
-                return responseError(HttpStatusCode.BadRequest, ex.toString())
-            })
+        if (name == null || whitelist == null || pubkey == null) {
+            return responseError(HttpStatusCode.BadRequest, reason)
+        } else {
```
Comment (dolgopolovwork) [inline src/main/kotlin/registration/RegistrationServiceEndpoint.kt]: Seems like there is no need in 'else' node. We can omit this.

File: src/integration-test/kotlin/integration/eth/WithdrawalIntegrationTest.kt [hunk 13]
```diff
@@ -94,7 +94,8 @@ class WithdrawalIntegrationTest {
                     "0x0000000000000000000000000000000000000000",
                     decimalAmount,
                     ethWallet,
-                    hash
+                    hash,
+                    ""
```
Comment (luckychess) [inline src/integration-test/kotlin/integration/eth/WithdrawalIntegrationTest.kt]: I guess a valid relay address is necessary here.

File: src/integration-test/kotlin/integration/eth/WithdrawalPipelineIntegrationTest.kt [hunk 14]
```diff
@@ -184,7 +184,7 @@ class WithdrawalPipelineIntegrationTest {
      */
     @Test
     fun testWithdrawInWhitelist() {
-        integrationHelper.registerClient(clientName, keypair)
+        integrationHelper.registerClient(clientName, emptyList(),keypair)
```
Comment (luckychess) [inline src/integration-test/kotlin/integration/eth/WithdrawalPipelineIntegrationTest.kt]: Formatting please.

File: src/integration-test/kotlin/integration/helper/IntegrationHelperUtil.kt [hunk 15]
```diff
@@ -211,7 +213,7 @@ class IntegrationHelperUtil {
     fun registerBtcAddress(irohaAccountName: String): String {
         val keypair = ModelCrypto().generateKeypair()
         preGenBtcAddress().fold({
-            btcRegistrationStrategy.register(irohaAccountName, keypair.publicKey().hex())
+            btcRegistrationStrategy.register(irohaAccountName, emptyList(),keypair.publicKey().hex())
```
Comment (luckychess) [inline src/integration-test/kotlin/integration/helper/IntegrationHelperUtil.kt]: Formatting.

File: src/integration-test/resources/eth/registration.properties [hunk 16]
```diff
@@ -16,3 +16,15 @@ eth-registration.iroha.creator=eth_registration_service_red@notary
...
+eth-notary.ethereum.url=http://d3-eth-node0:8545
+eth-notary.ethereum.confirmationPeriod=0
+# path to Ethereum keys
+eth-notary.ethereum.credentialsPath=deploy/ethereum/keys/ganache.key
+# Gas price for relay contract deployment
+eth-notary.ethereum.gasPrice=1
+# Gas limit for relay contract deployment
+eth-notary.ethereum.gasLimit=1999999
+# Relay registry contract address
+eth-registration.ethRelayRegistryAddress=0x481776924972f1b25d13ad3820e6f01d15757002
```
Comment (luckychess) [inline src/integration-test/resources/eth/registration.properties]: Looks like it doesn't work for ganache environment.

File: src/integration-test/resources/eth/registration_deploy.properties [hunk 17]
```diff
@@ -16,3 +16,15 @@ eth-registration.iroha.creator=eth_registration_service_red@notary
...
+eth-notary.ethereum.url=http://d3-eth-node0:8545
+eth-notary.ethereum.confirmationPeriod=0
+# path to Ethereum keys
+eth-notary.ethereum.credentialsPath=deploy/ethereum/keys/ganache.key
+# Gas price for relay contract deployment
+eth-notary.ethereum.gasPrice=1
+# Gas limit for relay contract deployment
+eth-notary.ethereum.gasLimit=1999999
+# Relay registry contract address
+eth-registration.ethRelayRegistryAddress=0x481776924972f1b25d13ad3820e6f01d15757002
```
Comment (luckychess) [inline src/integration-test/resources/eth/registration_deploy.properties]: Same here.

File: src/main/kotlin/registration/IrohaAccountCreator.kt [hunk 18]
```diff
@@ -29,31 +30,50 @@ class IrohaAccountCreator(
      */
     fun create(
         currencyAddress: String,
+        whitelist: String,
```
Comment (luckychess) [inline src/main/kotlin/registration/IrohaAccountCreator.kt]: Please update function description as well.

File: src/main/resources/eth/notary_deploy.properties [hunk 19]
```diff
@@ -23,7 +23,7 @@ eth-notary.ethereum.confirmationPeriod=1
 # path to Ethereum keys
 eth-notary.ethereum.credentialsPath=deploy/ethereum/keys/ropsten.key
 # Gas price for relay contract deployment
-eth-notary.ethereum.gasPrice=1
+eth-notary.ethereum.gasPrice=1000000000
```
Comment (luckychess) [inline src/main/resources/eth/notary_deploy.properties]: Is it important for ropsten?

File: src/main/kotlin/sidechain/eth/util/Crypto.kt [hunk 20]
```diff
@@ -34,14 +34,16 @@ fun signUserData(ecKeyPair: ECKeyPair, toSign: String): String {
...
  * @return keccak-256 hash of all provided fields
  */
-fun hashToWithdraw(tokenAddress: String, amount: AmountType, accountAddress: String, irohaHash: String): String {
+fun hashToWithdraw(tokenAddress: String, amount: AmountType, accountAddress: String, irohaHash: String, from: String): String {
     return Hash.sha3(
         tokenAddress.replace("0x", "")
                 + String.format("%064x", BigInteger(amount)).replace("0x", "")
                 + accountAddress.replace("0x", "")
                 + irohaHash.replace("0x", "")
+                + from.replace("0x","")
```
Comment (luckychess) [inline src/main/kotlin/sidechain/eth/util/Crypto.kt]: Space.

File: src/main/kotlin/registration/eth/main.kt [hunk 21]
```diff
@@ -16,13 +18,15 @@ private val logger = KLogging().logger
 fun main(args: Array<String>) {
     val registrationConfig =
         loadConfigs("eth-registration", EthRegistrationConfig::class.java, "/eth/registration.properties")
-    executeRegistration(registrationConfig)
+    val passwordConfig = loadEthPasswords("eth-notary", "/eth/ethereum_password.properties", args)
```
Comment (luckychess) [inline src/main/kotlin/registration/eth/main.kt]: I guess registration service should use its own prefix (e.g. "eth-registration").

File: src/main/resources/eth/registration_deploy.properties [hunk 22]
```diff
@@ -16,3 +16,14 @@ eth-registration.iroha.creator=eth_registration_service_red@notary
...
+# URL of Ethereum client
+eth-registration.ethereum.url=http://d3-eth-node0:8545
+eth-registration.ethereum.confirmationPeriod=0
+# path to Ethereum keys
+eth-registration.ethereum.credentialsPath=deploy/ethereum/keys/ropsten.key
+# Gas price for relay contract deployment
+eth-registration.ethereum.gasPrice=1000000000
+# Gas limit for relay contract deployment
+eth-registration.ethereum.gasLimit=1999999
+eth-registration.ethRelayRegistryAddress=0x70c0f7fd61277814b0c03425b6a92ba003adbc9b
```
Comment (luckychess) [inline src/main/resources/eth/registration_deploy.properties]: Same contract should be deployed for mainnet as well.

File: src/main/resources/eth/relay_registration_deploy.properties [hunk 23]
```diff
@@ -4,7 +4,7 @@ relay-registration.number=10
 # Account to store registered free wallets
 relay-registration.notaryIrohaAccount=notary_red@notary
 # Master smart contract address
-relay-registration.ethMasterWallet=0x57dd50c6b4ec7a33ee3bb5056bf38856e181684c
+relay-registration.ethMasterWallet=0xde2173601b9c44f3ee4af212186902e5986bde5a
```
Comment (luckychess) [inline src/main/resources/eth/relay_registration_deploy.properties]: Same contract should be deployed for mainnet as well.

**Pull request merged**